pyface/data_view/tests/test_data_view_widget.py

```diff
@@ -185,6 +502,52 @@ def test_selection_mode_extended(self):
             [((1, 4), ()), ((2, 0), ())],
         )
 
+    def test_selection_mode_change_when_selection_change(self):
+        modes = cycle(["extended", "single"])
+
+        self.widget.selection_mode = next(modes)
+        self._create_widget_control()
+        self.assertEqual(self.widget._get_control_selection_mode(), "extended")
+
+        def change_selection_mode(event):
+            if self.widget.selection:
+                self.widget.selection_mode = next(modes)
+
+        self.widget.observe(
+            change_selection_mode, "selection.items", dispatch="ui")
+        self.addCleanup(
+            self.widget.observe,
+            change_selection_mode, "selection.items", dispatch="ui",
+            remove=True,
+        )
+
+        # switch from "extended" to "single"
+        self.widget.selection = [((0,), ()), ((1,), ())]
+        self.gui.process_events()
+        self.assertEqual(self.widget._get_control_selection_mode(), "single")
+
+        if is_wx:
+            # Not sure why the selection is not updated.
+            self.assertEqual(
+                self.widget._get_control_selection(), [((0,), ()), ((1,), ())]
+            )
```

Comment:
The expected value is an empty list. I tried to understand why this did not work for wx, but I couldn't.

Reply:
Weirdly, this assertion fails on Travis CI OSX build, but passes on Travis CI Linux build.
The assertion passes (i.e. expected failure) on my local machine too... It'd be good to get an independent confirmation. :puzzled:

Reply:
On Travis CI linux build, it printed this error message and the test continued to "pass"(?):
```
test_selection_mode_change_when_selection_change (pyface.data_view.tests.test_data_view_widget.TestWidget) ... Traceback (most recent call last):
  File "/Users/travis/.edm/envs/pyface-test-3.6-wx/lib/python3.6/site-packages/pyface-7.1.0.dev110-py3.6.egg/pyface/data_view/i_data_view_widget.py", line 247, in _update_selection
    self.selection = new
  File "/Users/travis/.edm/envs/pyface-test-3.6-wx/lib/python3.6/site-packages/pyface-7.1.0.dev110-py3.6.egg/pyface/data_view/i_data_view_widget.py", line 127, in validate
    "is 'single', got {!r}".format(value)
traits.trait_errors.TraitError: Selection must have at most one element when selection_mode is 'single', got [((0,), ()), ((1,), ())]
Traceback (most recent call last):
  File "/Users/travis/.edm/envs/pyface-test-3.6-wx/lib/python3.6/site-packages/pyface-7.1.0.dev110-py3.6.egg/pyface/data_view/i_data_view_widget.py", line 247, in _update_selection
    self.selection = new
  File "/Users/travis/.edm/envs/pyface-test-3.6-wx/lib/python3.6/site-packages/pyface-7.1.0.dev110-py3.6.egg/pyface/data_view/i_data_view_widget.py", line 127, in validate
    "is 'single', got {!r}".format(value)
traits.trait_errors.TraitError: Selection must have at most one element when selection_mode is 'single', got [((0,), ()), ((1,), ())]
ok
```